guides/inventing-service.md

```diff
@@ -0,0 +1,856 @@
+
+Next-up you might want to add supports for timeouts. A timeout is the max
+duration `handler` is allowed to take. If it doesn't produce a response within
+that amount of time an error is returned.
+
+Your first idea might be to modify `Server` to support being configured with a
+timeout which it will somehow apply to the call of the handler. However it turns out
+you can actually add a timeout without modifying `Server`. Using
+`tokio::time::timeout` we can make a new handler function that calls our
+previous `handle_request` but with a timeout of 30 seconds:
+
+```rust
+async fn handler_with_timeout(request: Request) -> Result<Response, Error> {
+    let result = tokio::time::timeout(
+        Duration::from_secs(30),
+        handle_request(request)
+    ).await;
+
+    match result {
+        Ok(Ok(response)) => Ok(response),
+        Ok(Err(error)) => Err(error),
+        Err(_timeout_elapsed) => Err(Error::timeout()),
+    }
+}
+```
+
+This provides a quite nice separation of concerns. We were able to add a timeout
+without changing any of our existing code.
+
+Lets add one more feature in this way. Imagine we're building a JSON API and
+would therefore like a `Content-Type: application/json` header on all responses.
+We can wrap `handler_with_timeout` in a similar way and modify the response like
+so:
+
+```rust
+async fn handler_with_timeout_and_content_type(
+    request: Request,
+) -> Result<Response, Error> {
+    let mut response = handler_with_timeout(request).await?;
+    response.set_header("Content-Type", "application/json");
+    Ok(response)
+}
+```
+
+We now have a handler that will process an HTTP request, take no longer than 30
+seconds, and always have the right `Content-Type` header all without modifying
+our original `handle_request` function or `Server` struct.
+
+Designing libraries that can be extended in this way is very powerful since it
+allows users to extend things without having to wait for the library maintainers
+to add support for it.
+
+However there is one problem. Our current setup doesn't scale very well. Imagine
+we have many `handle_with_*` functions that each add a little bit of behavior.
+Having to hard code the chain of which intermediate handler calls which isn't
+great. Our current chain is
+
+1. `handler_with_timeout_and_content_type` which calls
+2. `handler_with_timeout` which calls
+3. `handle_request` which actually processes the request
+
+It would be nice if we could somehow compose these three functions without
+having to hard code the exact order. Something like:
+
+```rust
+let final_handler = with_content_type(with_timeout(handle_request));
+```
+
+While still being able to run our handler like before:
+
+```rust
+server.run(final_handler).await?;
+```
+
+You could imagine `with_content_type` and `with_timeout` being functions that
+took an argument of type `F: Fn(Request) -> Future<Output = Response>` and
+returned a closure a la `impl Fn(Request) -> Future<Output = Result<Response,
```

Comment:
Hey, does the word `la` is a typo?

Reply:
Nope:

> Borrowed from French, à la means "according to" or "in the manner of,"

https://www.dictionary.com/e/slang/a-la/

I have changed it however to "like" as that is probably more understandable.